File: CDSS/src/main/java/sbnz/ftn/uns/ac/rs/cdss/services/PatientService.java
package sbnz.ftn.uns.ac.rs.cdss.services;

import java.util.Collection;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import sbnz.ftn.uns.ac.rs.cdss.model.Patient;
import sbnz.ftn.uns.ac.rs.cdss.model.dto.AlergiesDetailsDTO;
import sbnz.ftn.uns.ac.rs.cdss.model.dto.MedicineDetailsDTO;
import sbnz.ftn.uns.ac.rs.cdss.model.dto.MedicineRecordDTO;
import sbnz.ftn.uns.ac.rs.cdss.model.dto.PatientDTO;
import sbnz.ftn.uns.ac.rs.cdss.model.dto.PatientDetailsDTO;
import sbnz.ftn.uns.ac.rs.cdss.model.dto.ReportDTO;


public interface PatientService {

	Page<PatientDetailsDTO> getAllPatients(String username, Pageable pageable);
	AlergiesDetailsDTO getAlergies(String username,Long id);
	PatientDetailsDTO getById(String username, Long id);
	PatientDetailsDTO save(String username, PatientDTO patient);
	PatientDetailsDTO update(String username, PatientDTO patient, Long id);
	void deletePatient(String username, Long id);
	MedicineDetailsDTO addAlergie(String username, Long id, String name);
	
	Collection<ReportDTO> getReport1(String username);
	Collection<ReportDTO> getReport2(String username);
	Collection<ReportDTO> getReport3(String username);

	Page<PatientDetailsDTO> getByFN(String username,String firstname, Pageable pageable);
	Page<PatientDetailsDTO> getByLN(String username,String lastname, Pageable pageable);
	Page<PatientDetailsDTO> getByFNLN(String username,String firstname,String lastname, Pageable pageable);

	Collection<MedicineRecordDTO> getMR(String username, Long id);
}

File: CDSS-client/src/app/services/patient/patient.service.ts
import { Injectable } from '@angular/core';
import {HttpClient, HttpParams} from '@angular/common/http';
import {Observable} from 'rxjs/Observable';
import {IPageable} from '../../model/IPageable';
import { IPatient } from '../../model/IPatient';
import { IMedicine } from '../../model/IMedicine';
import { IAlergie } from '../../model/IAlergie';
import { IReport } from '../../model/IReport';
import { IMedicalRecord } from '../../model/IMedicalRecord';

@Injectable()
export class PatientService {

  constructor(private http: HttpClient) { }

  add(patient: IPatient): Observable<IPatient> {
    const urlPath = '/api/patients';
    return this.http.post<IPatient>(urlPath, patient);
  }

  addAlergie(id: any, name:String): Observable<IPatient> {
    const urlPath = '/api/patients/alergies/'+id+"/"+name;
    return this.http.put<IPatient>(urlPath,null);
  }

  update(patient: IPatient, id:any): Observable<IPatient> {
    const urlPath = '/api/patients/'+id;
    return this.http.put<IPatient>(urlPath, patient);
  }

  getReport1(): Observable<IReport[]> {
    const urlPath = '/api/patients/report1';
    return this.http.get<IReport[]>(urlPath);
  }

  getReport2(): Observable<IReport[]> {
    const urlPath = '/api/patients/report2';
    return this.http.get<IReport[]>(urlPath);
  }

  getReport3(): Observable<IReport[]> {
    const urlPath = '/api/patients/report3';
    return this.http.get<IReport[]>(urlPath);
  }
  getAll(page: number, limit: number): Observable<IPageable> {
    page = page - 1;
    const urlPath = '/api/patients?&page=' + page + '&size=' + limit;
    return this.http.get<IPageable>(urlPath);
  }

  getById(id:any): Observable<IPatient> {
    const urlPath = '/api/patients/'+id;
    return this.http.get<IPatient>(urlPath);
  }

  getAlergies(id:any): Observable<IAlergie> {
    const urlPath = '/api/patients/alergies/'+id;
    return this.http.get<IAlergie>(urlPath);
  }

  getMR(id:any): Observable<IMedicalRecord[]> {
    const urlPath = '/api/patients/mrecords/'+id;
    return this.http.get<IMedicalRecord[]>(urlPath);
  }

  getByFN(fn:String, page: number, limit: number): Observable<IPageable> {
    page = page - 1;
    let params = new HttpParams();
    params = params.append('firstname', String(fn));   
    params = params.append('page', String(page));
    params = params.append('size', String(limit));

    const urlPath = '/api/patients/filter';    
    return this.http.get<IPageable>(urlPath,{params:params});
  }

  getByLN(ln:String, page: number, limit: number): Observable<IPageable> {
    page = page - 1;
    let params = new HttpParams();
    params = params.append('lastname',String(ln));    
    params = params.append('page', String(page));
    params = params.append('size', String(limit));
    
    const urlPath = '/api/patients/filter';    
    return this.http.get<IPageable>(urlPath,{params:params});
  }

  getByFNLN(fn:String,ln:String,  page: number, limit: number): Observable<IPageable> {
    page = page - 1;
    let params = new HttpParams();
    params = params.append('firstname', String(fn));   
    params = params.append('lastname',String(ln));    
    params = params.append('page', String(page));
    params = params.append('size', String(limit));
    
    const urlPath = '/api/patients/filter';    
    return this.http.get<IPageable>(urlPath,{params:params});
  }


}
